### DISCUSSION TITLE
Add texture on text with dynamic size

### DISCUSSION THREAD

**Discussion by pangong** [ImageMagick/ImageMagick#2084] [Help]
Hi,

I am trying to create text with texture (using `-tile`).
Everything is fine when i use `-annotate` with fixed `-pointsize`.

`convert -size 800x65 xc:none -font my-font.otf -gravity center -tile texture.jpg -pointsize 52 -annotate +0+0 "Some test string" output.png`

But i have fixed width of image and i would like to not specify pointsize for longer strings.
Is there any possibility to add texture and not using annotate.

When i am using `label:`, `-tile` is ignored and string has no texture. 

`convert -size 800x65 -font my-font.otf -gravity center -tile texture.jpg -background transparent label:'My looooooooooong looong text sd fsd sd f sd' output.png`

Please help.

**Comment by fmw42** [comment]
One way is to create your label: text image with transparent background, then clone it and put texture in the clone. Then composite the text image over the texture image.

Texture:

![tile_aqua](https://user-images.githubusercontent.com/13458218/83434625-afca3f80-a3ef-11ea-96d6-fa90a9adce7d.jpg)

Unix Syntax:

```
convert -size 500x100 -background none -font candice \
-gravity center label:"This is some text" \
\( +clone -tile tile_aqua.jpg -draw "color 0,0 reset" \) \
+swap -gravity center -compose over -composite result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/83434683-c7092d00-a3ef-11ea-82ff-bb551e7b1646.jpg)

**Reply by pangong** [reply]
Hi,

Thanks for the answer.
But I would like to achieve this kind of effect without worrying about text size.

![output](https://user-images.githubusercontent.com/35529301/83440596-7c1a0480-a445-11ea-87f0-d353d87b6360.png)

I am doing it using this command:
`convert -size 500x100 xc:lightyellow -background blue -font candice.ttf \
-gravity center -tile tile_aqua.jpg  -pointsize 52 -annotate +0+0 "This is some text" output.png`

Pattern only on text.
The problem I found that `-annotate` doesn't know the size of text so it can be dynamic to fit the box. I am wondering if I can do it without checking each size until it fits.

**Reply by fmw42** [reply]
Create a white text on black background image and use it as a mask in the composite of a yellow image and a texture image.

![tile_aqua](https://user-images.githubusercontent.com/13458218/83442999-e6a75200-a3fd-11ea-8a14-9711527cec52.jpg)


```
convert -size 500x100 -background black -fill white -font candice -gravity center label:"This is some text" \
\( -clone 0 -tile tile_aqua.jpg -draw "color 0,0 reset" \) \
\( -clone 0 -fill yellow -colorize 100 \) \
-reverse -gravity center -compose over -composite result.jpg
```

<br>
Or perhaps better (black text on white background)

```
convert -size 500x100 -background white -fill black -font candice -gravity center label:"This is some text" \
\( -clone 0 -fill yellow -colorize 100 \) \
\( -clone 0 -tile tile_aqua.jpg -draw "color 0,0 reset" \) \
-reverse -gravity center -compose over -composite result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/83442986-e018da80-a3fd-11ea-9647-cd0d52bff6c9.jpg)

**Reply by pangong** [reply]
I need to test masks. 
Your examples generate me only background wihtout text.

![result](https://user-images.githubusercontent.com/35529301/83444541-d5853200-a44b-11ea-9541-fa8ade34d1af.jpg)

![result](https://user-images.githubusercontent.com/35529301/83444596-e5047b00-a44b-11ea-9d9a-a9a939f592b6.jpg)

**Reply by fmw42** [reply]
What is your IM version and platform?  My example was for IM 6.x.x.x and Unix syntax. If you are on Windows or IM 7, then changes are needed. Always provide that information when asking questions on this forum.

**Reply by pangong** [reply]
Sorry for that.
I have IM 7 on osx.

**Reply by fmw42** [reply]
This should work:

```
magick -size 500x100 -background white -fill black -font candice -gravity center label:"This is some text" \
\( -clone 0 -fill yellow -colorize 100 \) \
\( -clone 0 -tile tile_aqua.jpg -draw "color 0,0 reset" \) \
-reverse -gravity center -compose over -composite result.jpg
```

<br>
Note that the end of line backslashes "\" are important and should not have any spaces after them. They signify the code is continued on the next line.  You can remove the backslashes and make the command all one long line as


```
magick -size 500x100 -background white -fill black -font candice -gravity center label:"This is some text" \( -clone 0 -fill yellow -colorize 100 \) \( -clone 0 -tile tile_aqua.jpg -draw "color 0,0 reset" \) -reverse -gravity center -compose over -composite result.jpg`
```

**Comment by GeeMack** [comment]
If you're using IMv7 you can get dynamically sized, tile-filled text using a command like this (using IM's built-in "rose:" to make the tile)....

     magick rose: -write mpr:rosetile +delete \
        -size 720x96 -gravity center -background none -fill none \
        label:"This is it." -pointsize "%[label:pointsize]" \
        -fill mpr:rosetile -annotate 0,0 "This is it." tiledtext.png

That starts by writing the tile image to a temporary ImageMagick memory register "mpr:rosetile". Then it specifies the desired dimensions for the label. Next it creates the label with a transparent fill on a transparent background. So you have a canvass the size you want.

Now, with IMv7 you get access to a variable which holds the pointsize of the most recent dimension-fitted label. So you set the pointsize to that variable with "-pointsize %[label:pointsize]".

After that you can set the background to whatever you want, set the fill to the tile image with "-fill mpr:rosetile", and  "-annotate" or "-draw" will use the correct pointsize to create the tile filled text on the specified size canvas.
